Frontend/src/Components/Graph/CustomGraph.js

```diff
@@ -1,427 +1,496 @@
+  // called when the name is to be saved, memoized for performance
+  const onNameSave = useCallback(
+    (name, isNew) => onSave(name, isNew, datasetKeys),
+    [onSave, datasetKeys]
+  );
+  const [historyLength, setHistoryLength] = useState(
+    secondsRetained ?? GraphData.defaultHistoryLength
+  );
+
+  // side-effects for when new datasets are chosen
+  useEffect(() => {
+    // console.log("yo, new datasets dropped:", datasetKeys);
+    // console.log("keys:", keys);
+
+    // update dataset object {key: boolean}
+    const newDatasets = {};
+    for (const key of datasetKeys) {
+      newDatasets[key] = datasets[key] ?? true;
+    }
+    // console.log("new datasets:", newDatasets);
+    setDatasets(newDatasets);
+  }, [datasetKeys]);
+
+  useEffect(() => {
+    // update formatted datasets storage array if all datasets are updated
+    const formattedDatasets = [];
+    for (const key of datasetKeys) {
+      const temp = packedData.find(
+        (packedDataset) => packedDataset.key === key
+      );
+      if (temp) {
+        const copy = Object.assign({}, temp);
+        copy.hidden = !datasets[key];
+        // console.log(`${title} pushed:`, copy);
+        formattedDatasets.push(copy);
+      } else {
+        console.warn(`unable to find "${key}" for "${title}"`);
+      }
+    }
+
+    // console.log("formatted datasets:", formattedDatasets);
+    setFormattedDatasets(formattedDatasets);
+  }, [datasets, packedData]);
+
+  // console.log("formatted datasets:", formattedDatasets);
+  // console.log(datasets, "keys:", datasetKeys);
+
+  // const onSelectSave = (keys) => {
+  //   // updateDatasets({ action: "set", key: keys })
+  //   // console.log("keys:", keys);
+  //   const newDatasets = {};
+  //   for (const key of keys) {
+  //     newDatasets[key] = datasets[key] ?? true;
+  //   }
+  //   // console.log("new datasets:", newDatasets);
+  //   setDatasets(newDatasets);
+  // };
+
+  // the modal that appears when the user wants to update the datasets that are shown
+  const graphSelectModal = useMemo(() => {
+    const onSelectSave = (newDatasetKeys, newHistoryLength) => {
+      setDatasetKeys(newDatasetKeys);
+      setHistoryLength(newHistoryLength);
+    };
+    return (
+      <GraphSelectModal
+        isOpen={isDataSelectOpen}
+        onClose={onDataSelectClose}
+        initialDatasets={datasetKeys}
+        initialHistoryLength={historyLength}
+        onSave={onSelectSave}
+      />
+    );
+  }, [
+    isDataSelectOpen,
+    onDataSelectClose,
+    datasetKeys,
+    setDatasetKeys,
+    setHistoryLength,
+    historyLength,
+  ]);
+
+  // the modal that appears when the user saves the graph,
+  // but the graph has no name associated with it
+  const graphNameModal = useMemo(
+    () => (
+      <GraphNameModal
+        isOpen={isNameModalOpen}
+        onClose={onNameModalClose}
+        onSave={onNameSave}
+      />
+    ),
+    [isNameModalOpen, onNameModalClose, onNameSave]
```

Comment:
Same question here. Why onNameSave?

Reply:
Same reason; changing the callback would change the resulting component.